### PR TITLE
Test for tcms.testplans.views.Edit. Refs #1617

### PR DESCRIPTION
**Pull request by cmbahadir**
Added tests for tcms.testplans.views.Edit.

### CODE REVIEW

File: tcms/testplans/tests/test_edit_plan.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            'email_settings-MIN_NUM_FORMS': '0',
+            'email_settings-MAX_NUM_FORMS': '1',
+        }
+
+        user_should_have_perm(cls.tester, perm='testplans.change_testplan')
+
+    def test_404_if_testplan_id_not_exist(self):
+        url = reverse('plan-edit', args=[99999])
+        response = self.client.get(url)
+        self.assertEqual(HTTPStatus.NOT_FOUND, response.status_code)
```
Comment (atodorov) [inline tcms/testplans/tests/test_edit_plan.py:42]: This functionality is provided by the underlying UpdateView and we don't have to test it explicitly.

File: tcms/testplans/tests/test_edit_plan.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    def test_show_testplan_edit_page(self):
+        response = self.client.get(self.testplan_edit_url)
+        self.assertEqual(HTTPStatus.OK, response.status_code)
+        self.assertContains(response, self.plan.name)
```
Comment (atodorov) [inline tcms/testplans/tests/test_edit_plan.py:47]: You can assert a bit more things here. Like for example are we on the Edit Test plan page indeed! (see page title)

Comment (atodorov) [reply tcms/testplans/tests/test_edit_plan.py:47]: Also the `notify_formset` in `context` is something that is important and could break so an assert that all of the values are as expected will verify everything is fine in that part of the code.

File: tcms/testplans/tests/test_edit_plan.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                            'name="email_settings-TOTAL_FORMS" value="1"')
+
+    def test_edit_testplan_text_field(self):
+        edit_data = self.testplan_edit_data.copy()
+        new_text = 'Edited: {0}'.format(self.plan.text)
+        edit_data['text'] = new_text
+
+        response = self.client.post(self.testplan_edit_url, data=edit_data)
+
+        self.assertEqual(HTTPStatus.FOUND, response.status_code)
```
Comment (atodorov) [inline tcms/testplans/tests/test_edit_plan.py:63]: Use `follow=True` for `client.post()` and you should land on the TP page with a 200 response. `assertContains` checks both the response code & the text that you specify.

File: tcms/testplans/tests/test_edit_plan.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        self.assertContains(response,
+                            'name="email_settings-TOTAL_FORMS" value="1"')
+        self.assertContains(response,
+                            'name="email_settings-INITIAL_FORMS" value="1"')
+        self.assertContains(response,
+                            'name="email_settings-MIN_NUM_FORMS" value="0"')
+        self.assertContains(response,
+                            'name="email_settings-MAX_NUM_FORMS" value="1"')
+        self.assertContains(response,
+                            'name="email_settings-TOTAL_FORMS" value="1"')
```
Comment (atodorov) [inline tcms/testplans/tests/test_edit_plan.py:54]: These asserts should check the values of the on/off buttons, they are more important than these hidden fields, see previous comment.

File: tcms/testplans/tests/test_edit_plan.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        edit_data['text'] = new_text
+
+        response = self.client.post(self.testplan_edit_url, data=edit_data, follow=True)
+
+        self.assertContains(response, new_text)
+        self.plan.refresh_from_db()
+        self.assertEqual(new_text, self.plan.text)
+
+    def test_try_editting_testplan_with_invalid_form(self):
+        testplan = TestPlanFactory(
```
Comment (cmbahadir) [inline tcms/testplans/tests/test_edit_plan.py:62]: Added a invalid form testcase specific `testplan (id=2)` object, otherwise super classes `testplan (id=1)` instance objects `cleaned_data` overrides the invalid field(s) here;
https://github.com/kiwitcms/Kiwi/blob/508564d0fd90bcdab725684c2804126f5092f0b0/tcms/testplans/views.py#L82

which makes the form valid and it continues again with the valid form branch:
https://github.com/kiwitcms/Kiwi/blob/508564d0fd90bcdab725684c2804126f5092f0b0/tcms/testplans/views.py#L83

**Comment by cmbahadir** [comment]
Refactor in progress , please do not review.

File: tcms/testplans/tests/test_views.py [hunk 6]
```diff
@@ -54,3 +56,167 @@ def test_view_tags_without_permissions(self):
         # assert tag actions are shown
         self.assertNotContains(response, 'Add Tag')
         self.assertContains(response, '<span class="disabled grey">%s</span>' % _('Remove'))
+
+
+class NewPlanViewTest(BasePlanTest):
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:61]: This is unrelated so it was better to leave it as it was. If you revert it that would be better.

File: tcms/testplans/tests/test_views.py [hunk 7]
```diff
@@ -54,3 +56,167 @@ def test_view_tags_without_permissions(self):
...
+        response = self.client.get(self.location, follow=True)
+        self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
+
+    def test_post_with_no_perm_redirects_to_login(self):
+        remove_perm_from_user(self.user, self.add_testplan_permission)
+        response = self.client.post(self.location, self.request, follow=True)
+        self.assertRedirects(response, reverse('tcms-login') + '?next=' + self.location)
+
+
+class EditPlanViewTest(BasePlanTest):
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:134]: BasePlanTest looks to me like some old test class which we've added long ago. It's better to use LoggedInTestCase directly (less levels of abstractions and less shared data). You will only need to create a TP object for editting.

File: tcms/testplans/tests/test_views.py [hunk 8]
```diff
@@ -54,3 +56,167 @@ def test_view_tags_without_permissions(self):
...
+        testplan_edit_url = reverse('plan-edit', args=[test_plan.pk])
+        edit_data = self.testplan_edit_data.copy()
+        edit_data['email_settings-0-auto_to_case_owner'] = 'invalid_value'
+
+        response = self.client.post(testplan_edit_url, data=edit_data, follow=True)
+
+        assert_notify_buttons(self, response)
+
+
+def assert_notify_buttons(self, response):
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:198]: Indentation!. This should be a class method, not a function.

File: tcms/testplans/tests/test_views.py [hunk 9]
```diff
@@ -54,3 +56,167 @@ def test_view_tags_without_permissions(self):
...
+
+    def test_try_editting_testplan_with_invalid_form(self):
+        test_plan = TestPlanFactory(
+            author=self.user,
+            product=self.product,
+            product_version=self.product_version)
+
+        testplan_edit_url = reverse('plan-edit', args=[test_plan.pk])
+        edit_data = self.testplan_edit_data.copy()
+        edit_data['email_settings-0-auto_to_case_owner'] = 'invalid_value'
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:191]: Hmmm, I'm not sure how that works in practices. We've never tried setting these fields to something different that on/off. In any case if you want to use this there must be an assert that this button shows an error message.

Another possibility is to set the `type` field to an invalid ID (e.g. -1) and there should be an error message - you can modify the value for the select field in the browser to play with it. Needs assert that there is an error message as well.

File: tcms/testplans/tests/test_views.py [hunk 10]
```diff
@@ -54,3 +56,167 @@ def test_view_tags_without_permissions(self):
...
+
+        response = self.client.post(testplan_edit_url, data=edit_data, follow=True)
+
+        assert_notify_buttons(self, response)
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:195]: Now this doesn't properly assert the state after an error on the page.  B/c these buttons default to True/On you should change some of them to False/off in the post data and after the assert about the error message also assert that the buttons which were switched off are still off and the rest are still on. 

With this you are trying to verify this line of code:
```context['notify_formset'] = kwargs.get('notify_formset') or PlanNotifyFormSet()```

If we've got a bound form (even if it has errors) the page will show the error message and will preserve the button choices which have already been selected previously. If you remove the left part between `=` and `or` button choices will not be remembered between form submissions when there are errors.

File: tcms/testplans/tests/test_views.py (hunk 9 above)
Comment (cmbahadir) [reply tcms/testplans/tests/test_views.py:191]: Changing type to an invalid id does not cover here:

https://github.com/kiwitcms/Kiwi/blob/508564d0fd90bcdab725684c2804126f5092f0b0/tcms/testplans/views.py#L82

Since the validation here made for `notify_formset`. So i would suggest to keep it invalid over `email_settings-0-auto_to_case_owner`.

For error message check an additional case would be possible.

File: tcms/testplans/tests/test_views.py (hunk 10 above)
Comment (cmbahadir) [reply tcms/testplans/tests/test_views.py:195]: Are you sure about this, because form will not be submitted when there is an error, and i would expect these values to not change?

Comment (atodorov) [reply tcms/testplans/tests/test_views.py:195]: Forms are always submitted to the backend view, even if they contain errors in their values. That's the whole point of forms, isn't it.

Perhaps you are thinking about 1 scenario and I am thinking about another one where both involve invalid forms.

File: tcms/testplans/tests/test_views.py (hunk 9 above)
Comment (atodorov) [reply tcms/testplans/tests/test_views.py:191]: Ah right, you want to cover the branch on line 83 and I've been thinking about the `or` expression in `get_context_data()` which is completely different thing.  Since there are 2 forms on this page - the notify one and the main one choose better names for the test methods and we won't get confused.

File: tcms/testplans/tests/test_views.py (hunk 10 above)
Comment (cmbahadir) [reply tcms/testplans/tests/test_views.py:195]: **Case 1 - When there is an invalid value on the main form:**
`notify_formset` is not preserved, except `is_active`. Its value is preserved, probably because it is also a member of the main_form.

I am not sure if this is an intended behaviour but when there is an invalid main form,  `kwargs.get('notify_formset') ` returns None and it reconstructs the `notify_formset` from `PlanNotifyFormSet(instance=self.object)` with default on/checked values.

**Case 2 - When there is an invalid value on notify form:**
`notify_formset` comes with the invalid value until here but on `rendered_content` after `render_to_response` it returns back to the default value, where all the notify values are on(checked).

`return self.render_to_response(self.get_context_data(notify_formset=notify_formset))`

A similar behavour also exist on NewTestPlan class.

vokoscreen-2020-05-26_14-00-18

IMO, asserting to default notify formset values is sufficient.

File: tcms/testplans/tests/test_views.py [hunk 11]
```diff
@@ -54,3 +55,127 @@ def test_view_tags_without_permissions(self):
...
+        self.assertEqual(new_text, self.test_plan_1.text)
+
+    def test_try_editting_testplan_with_invalid_notify_form(self):
+        test_plan = TestPlanFactory(
+            author=self.tester)
+
+        testplan_edit_url = reverse('plan-edit', args=[test_plan.pk])
+        edit_data = self.testplan_edit_data.copy()
+        edit_data['email_settings-0-auto_to_plan_author'] = 'invalid_value'
+        edit_data['is_active'] = False
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:122]: is_active is not part of the notify form, but is a field on the TP model (main form is automatically generated from that). Not really necessary in this scenario.

File: tcms/testplans/tests/test_views.py [hunk 12]
```diff
@@ -54,3 +55,127 @@ def test_view_tags_without_permissions(self):
...
+
+        response = self.client.post(testplan_edit_url, data=edit_data, follow=True)
+        self.assert_notify_form(response)
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:125]: I'm not really sure what's the expected behavior here. I can't reproduce it manually but if one of these ON/OFF buttons is invalid I would expect an error and not the default value to be present. 

But I'm not entirely certain how the code underneath behaves. Could be a bug there.

File: tcms/testplans/tests/test_views.py (hunk 10 above)
Comment (atodorov) [reply tcms/testplans/tests/test_views.py]: > A similar behavour also exist on NewTestPlan class.

It is the same underlying form and logic, that's why.

File: tcms/testplans/tests/test_views.py (hunk 12 above)
Comment (atodorov) [reply tcms/testplans/tests/test_views.py:125]: I see your comment here:

> Case 2 - When there is an invalid value on notify form:
> notify_formset comes with the invalid value on rendered_content after render_to_response it returns back to the  default value, where all the notify values are on(checked).

so I'm inclined to mark this as resolved. But you really don't need is_active here.

File: tcms/testplans/tests/test_views.py (hunk 10 above)
Comment (atodorov) [reply tcms/testplans/tests/test_views.py]: For Case 1 - there is a bug in the code.

You need the following patch:
```
@@ -85,7 +87,11 @@ class Edit(UpdateView):
             return super().form_valid(form)
 
         # taken from FormMixin.form_invalid()
-        return self.render_to_response(self.get_context_data(notify_formset=notify_formset))
+        return self.render_to_response(self.get_context_data(form=form, notify_formset=notify_formset))
+
+    def form_invalid(self, form):
+        notify_formset = PlanNotifyFormSet(self.request.POST, instance=self.object)
+        return self.render_to_response(self.get_context_data(form=form, notify_formset=notify_formset))
```

When the main form is invalid we don't execute the `form_valid()` method at all so nobody tells `get_context` that there is a second form and that's why it reconstructs it without keeping the existing selections.

The same goes for `form_valid` - if the notify_form happens to be invalid we show the page again, this time keeping the notify_form but not the main one. You already have a scenario for this so before applying the patch if you try to assert one of the fields from the main form if should fail (I expect the value to be what comes from the DB, not what the user entered). You are doing something like that for the is_active field but strangely enough it didn't fail.

File: tcms/testplans/tests/test_views.py (hunk 12 above)
Comment (atodorov) [reply tcms/testplans/tests/test_views.py:125]: I tried explicitly modifying the underlying field for notify_form inside the view in order to make it invalid. The cleaned result still came out as True!.

This is probably some quirk in Django internals which I don't have more time to spend figuring out. Looks like these boolean fields do not fail on invalid values but tolerate them. 

Please add a comment in this test scenario to indicate that this is the case and that's why all of the buttons are still ON. And this scenario is done.

File: tcms/testplans/tests/test_views.py (hunk 10 above)
Comment (atodorov) [reply tcms/testplans/tests/test_views.py]: > But tests have passed, both before and after applying the patch.

Yes because the current tests do not assert on the states of the on/off buttons when there is an error in the main form. Moreover to make sure that the patch really works at least one of these buttons should be OFF so that there is an expected difference from the default behavior.

Comment (cmbahadir) [reply tcms/testplans/tests/test_views.py]: I am not sure if i am doing everyhting right, but even i apply the patch, the returned content is the default response with default 'on' values. By the way i have already defined the notify_formset values as 'off' by default on SetUpTestData - edit_data.

image

I suspect from render_to_response method. Seems like everything returns to its default 'on' value when its rendered.

Comment (atodorov) [reply tcms/testplans/tests/test_views.py]: I don't see the patch applied in your pull request.

Comment (cmbahadir) [reply tcms/testplans/tests/test_views.py]: Do you want me to include the patch in this PR? I thought that is going to be considered on a separate issue.

Comment (atodorov) [reply tcms/testplans/tests/test_views.py]: You can't write a correct test if the underlying code is not working correctly. Include the patch so we can move forward with the test.

File: tcms/testplans/tests/test_views.py [hunk 13]
```diff
@@ -54,3 +55,125 @@ def test_view_tags_without_permissions(self):
         # assert tag actions are shown
         self.assertNotContains(response, 'Add Tag')
         self.assertContains(response, '<span class="disabled grey">%s</span>' % _('Remove'))
+
+
+class EditPlanViewTest(LoggedInTestCase):
+    test_plan_name = 'plan name'
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:61]: Unused attribute, remove it.

File: tcms/testplans/tests/test_views.py [hunk 14]
```diff
@@ -54,3 +55,125 @@ def test_view_tags_without_permissions(self):
...
+
+        response = self.client.post(self.testplan_edit_url, data=edit_data, follow=True)
+
+        self.assertContains(
+            response,
+            _('Select a valid choice. That choice is not one of the available choices.'))
+
+    def test_with_invalid_version_shows_error(self):
+        edit_data = self.testplan_edit_data.copy()
+        edit_data['product_version'] = -1
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:151]: Here it is better to use another VersionFactory and create a version which exists (in DB) but is not related to the currently selected product. The mechanics of this is in the NewPlanForm.populate() method which restricts the valid values of Version based on the currently selected product.

File: tcms/testplans/tests/test_views.py [hunk 15]
```diff
@@ -54,3 +55,125 @@ def test_view_tags_without_permissions(self):
...
+
+        self.assertContains(response, new_text, html=True)
+        self.test_plan_1.refresh_from_db()
+        self.assertEqual(new_text, self.test_plan_1.text)
+
+    # Issue: When there is an invalid value on notify_formset,
+    # returned response is always the default (ON).
+    def test_with_invalid_notify_form_value(self):
+        test_plan = TestPlanFactory(
+            author=self.tester)
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:119]: I don't see the reason why you should use a different TP for this particular scenario. Using the existing one will work equally well.

File: tcms/testplans/tests/test_views.py [hunk 16]
```diff
@@ -54,3 +55,125 @@ def test_view_tags_without_permissions(self):
...
+
+    def test_with_invalid_type_shows_error(self):
+        edit_data = self.testplan_edit_data.copy()
+        edit_data['type'] = -1
+
+        response = self.client.post(self.testplan_edit_url, data=edit_data, follow=True)
+
+        self.assertContains(
+            response,
+            _('Select a valid choice. That choice is not one of the available choices.'))
```
Comment (atodorov) [inline tcms/testplans/tests/test_views.py:147]: This scenario must verify that all of the on/off buttons are as we expect.

To make sure the currently selected value has been kept between errors (e.g. that the patch works) one of the buttons in notify form should be set to a different value than its default.

File: tcms/testplans/tests/test_views.py (hunk 15 above)
Comment (cmbahadir) [reply tcms/testplans/tests/test_views.py:119]: See reply;
https://github.com/kiwitcms/Kiwi/pull/1677#discussion_r429549452

Comment (atodorov) [reply tcms/testplans/tests/test_views.py:117]: This comment is bonkers. There is absolutely no difference for the code path whether you have 1 or 2 or a 1000 objects in the database.  When the notify_formset is invalid it is passed directly to get_context_data() and then render_to_response(). 

This test should use the existing self.test_plan_1 variable as well.

Comment (atodorov) [reply tcms/testplans/tests/test_views.py:117]: I did check it and your explanation is incorrect. 

Add `print('++++ form_valid - notify_formset is NOT valid', notify_formset.errors)` in form_valid(), after the if condtition which checks notify_formset and you will get:

```
++++ form_valid - notify_formset is NOT valid [{'plan': ['Test plan email settings with this Plan already exists.']}]
```

The reason for this error is that `'email_settings-0-id': cls.test_plan_1.emailing.pk` points to the previous TP (let's call it TP-1) but you are trying to edit TP-2. The value `invalid_value` for the ON/OFF button doesn't have any effect in this case.

Which is strange but I've commented before that I've seen this behavior, https://github.com/kiwitcms/Kiwi/pull/1677#discussion_r432384387. Let me debug some more to see what is happening underneath Django. If we can't reliably set notify_formset fields to invalid values then there isn't much point in trying to implement this whole scenario.

Comment (atodorov) [reply tcms/testplans/tests/test_views.py:117]: Building on the previous comment:


```
class BooleanField(Field):
    def to_python(self, value):
        if isinstance(value, str) and value.lower() in ('false', '0'):
            value = False
        else:
            value = bool(value)
```

bool('invalid') -> True - this is the definition of how strings are evaluated in a boolean context in the Python language. The underlying form field relies on this conversion (with a few special cases b/c non empty strings are always True) and doesn't perform any other validation because it knows that it will get either True or False.

So it is not possible to make the notify_formset invalid by manipulating any of these boolean fields. 

You can however set `email_settings-0-id` to a different ID/non-existing ID and the formset will become invalid. This can also be seen in the browser, just add `value="-5"` for example in the HTML before pressing the Save button on this page.

With the current patch in views.py the page is reloaded and any buttons which have been manipulated (e.g. they were ON by default but were turned OFF by the user) are still preserved. 

That is the only way to test the path which deals with invalid notify_form. 

So the current implementation of this scenario is mostly on track but that seems to be by accident. Needs small adjustments.

Comment (cmbahadir) [reply tcms/testplans/tests/test_views.py:117]: Ahh ok, sorry for the inconvenience, that was my fault, i didn't dive into details. :)
